rfc-171-remove-legacy-browser-js-support.md

```diff
@@ -0,0 +1,128 @@
+
+## Problem
+
+This RFC is taking a broad approach to legacy browsers, but the main point of contention is likely to be Internet Explorer 11, the legacy browser with the most significant number of ongoing visitors.
+
+At the time of writing the service manual (which was last updated in 2022) says "You will still need to support Internet Explorer 11 if your analytics data shows at least 2% of your users arriving at the service's start page are using it." It [recommends testing](https://www.gov.uk/service-manual/technology/designing-for-different-browsers-and-devices) only in Edge of all the Microsoft browsers.
+
+The Design System team have updated `govuk-frontend` to use [ES modules in version 5](https://frontend.design-system.service.gov.uk/importing-css-assets-and-javascript/#javascript), released in 2023. [ES modules](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Modules) are not recognised by legacy browsers, notably Internet Explorer 11 and below. These browsers are no longer supported by Microsoft. We need to keep up to date with `govuk-frontend`, which means deciding how we approach JS support for IE11. The Design System has their own definitions for [browser support](https://frontend.design-system.service.gov.uk/browser-support/).
+
+The presence of any modern JS such as ES modules will cause errors for older browsers. This is because they will attempt to parse all JS before running it, encounter unfamiliar syntax, and error. No JS will run on older browsers once any modern JS is introduced, which means that any download of JS will be wasted.
+
+## Proposal
+
+The proposal is to embrace modern JS so that legacy browsers do not download any JS but see the non-JS version of GOV.UK. This should still provide a usable experience of the site as the site is built using the principle of progressive enhancement. This will also reduce page weight for those browsers and allow us to embrace modern JS techniques, leading to further code improvements in future.
+
+We should implement ES modules initially, with further work done later to potentially use other modern JS techniques. This will be done on a problem solving basis rather than upgrading everything - we will identify issues and look at how modern JS could solve them.
+
+Legacy browsers will not download any JS imported using the modules syntax. This would work as follows:
+
+- the HTML includes a JS file that imports code from other JS files using module includes
+- most browsers download and execute this code as normal, pulling in the code from the other JS files
+- legacy browsers download this one file but don't recognise the imports and therefore don't import anything else, resulting in a minimal JS download weight and a non-JS version of GOV.UK
+
+In practice this looks something like this.
+
+```HTML
+<script type="module" src="application.js"></script>
+```
+
+```JavaScript
+// application.js
+import './module1.js';
```

Comment:
Are we decided on whether we're going to go down a .js or .mjs route?

Reply:
No decision on this (or discussion as far as I'm aware) but I don't think `.mjs` offers anything particularly helpful? And if we were to rename all the files we'd have to update a lot of manifest files. Happy to do it if we need to, but I'm not currently suggesting we should.

Reply:
Sure, it sounds like this is no longer is part of the scope anyway with: https://github.com/alphagov/govuk-rfcs/pull/171#discussion_r1577945388

Probably one to learn from Design System about given they adopted that approach, I get confused about where the JS ecosystem is with this convention. Manifest files will likely depend on the extent of bundling done - not sure if they're used in all the potential ways Rails can use JS.